Added registration-parser.js

dalingk opened the pull request:
Probably doesn't work yet

dalingk committed:
Added registration-parser.js

Probably doesn't work yet

kyle-rader commented on server/webhooks/registration-parser.js:
```diff
@@ -0,0 +1,84 @@
+"use strict";
```
We don't need to use `'use-strict' in our es6 files :)

kyle-rader commented on server/webhooks/registration-parser.js:
```diff
@@ -0,0 +1,84 @@
+"use strict";
+import 'babel-register';
```
We don't need `babel-register` either thanks to Meteor's build setup.

kyle-rader commented on server/webhooks/registration-parser.js:
```diff
@@ -0,0 +1,84 @@
+"use strict";
+import 'babel-register';
+import _ from 'lodash'
```
- Semicolon after statement
- We want to avoid importing all of lodash if possible.  Instead use named imports: example if I only need the extend function (`_.extend`) then i would do
  
  ```
  import { extend } from 'lodash';
  ```

kyle-rader commented on server/webhooks/registration-parser.js:
```diff
@@ -0,0 +1,84 @@
+import _ from 'lodash'
+
+class cashnetJSON {
+    constructor(cnj) {
```
Variable naming: No abbreviations outside canonical iterators.
(Those being things like

```
i, j, k, m, n,
acc=accumulator // used a lot in lodash transformation or reduce functions
obj - object - also used a lot in mapping functions
```

kyle-rader commented on server/webhooks/registration-parser.js:
```diff
@@ -0,0 +1,84 @@
+import 'babel-register';
+import _ from 'lodash'
+
+class cashnetJSON {
```
`ClassNamesCapitalCamelCase`

kyle-rader commented on server/webhooks/registration-parser.js:
```diff
@@ -0,0 +1,84 @@
+
+class cashnetJSON {
+    constructor(cnj) {
+        let { itemcnt } = cnj;
```
Yay deconstruction!

kyle-rader commented on server/webhooks/registration-parser.js:
```diff
@@ -0,0 +1,84 @@
+    refs(cnj) {
+        let { itemcnt } = cnj;
+        let refArray = [];
+        _.forEach(_.range(1, _.toInteger(itemcnt) + 1), function (index) {
```
-  A note on iterating.  Always prefer `map` or `transform` if you can before `forEach` they are faster.
- `Array.push` is slow.
- This loop is the same as:

``` javascript
for(let i = 0; i < itemcnt; i++)
// even cleaner with lodash
_.times(itemcnt, (i) => { } )
```
- Faster to declare and instantiate your Array with a size at the same time.

kyle-rader commented on server/webhooks/registration-parser.js:
```diff
@@ -0,0 +1,84 @@
+        let refArray = [];
+        _.forEach(_.range(1, _.toInteger(itemcnt) + 1), function (index) {
+            refArray.push({});
+            refArray[index - 1]['amount'] = cnj[`amount${index}`];
```
Object keys that are not dynamic or requiring strings should just use dot notation
i.e. `SomeObject.thisKey`

kyle-rader commented on server/webhooks/registration-parser.js:
```diff
@@ -0,0 +1,84 @@
+        _.forEach(_.range(1, _.toInteger(itemcnt) + 1), function (index) {
+            refArray.push({});
+            refArray[index - 1]['amount'] = cnj[`amount${index}`];
+            refArray[index - 1]['qty'] = cnj[`qty${index}`];
```
We can clean the last 3 lines up as one object instantiation like this:

``` javascript
refArray[i] = {
  amount: cnj[`qty${index}`],
  qty: cnj[`qty${index}`],
};
```

kyle-rader commented on server/webhooks/registration-parser.js:
```diff
@@ -0,0 +1,84 @@
+            refArray[index - 1]['qty'] = cnj[`qty${index}`];
+        });
+        let refRE = /ref(\d{1,2})type(\d{1,2})/;
+        _.forEach(_.keys(cnj), function (value) {
```
Everything above applies here: See what cleanup you can do on your own :)

kyle-rader commented on server/webhooks/registration-parser.js:
```diff
@@ -0,0 +1,84 @@
+        return refArray;
+    }
+    getParticipants(refs) {
+        let participants = [];
```
Same thing here

kyle-rader commented on server/webhooks/registration-parser.js:
```diff
@@ -0,0 +1,84 @@
+    }
+    getTshirts(refs) {
+        let tshirts = [];
+        _.forEach(refs, function(object) {
```
Same thing here

kyle-rader commented on server/webhooks/registration-parser.js:
```diff
@@ -0,0 +1,84 @@
+        });
+        return tshirts;
+    }
+    getTotalAmount(refs) {
```
Use `reduce` (from `lodash`)

kyle-rader commented on server/webhooks/registration-parser.js:
```diff
@@ -0,0 +1,84 @@
+        return sum;
+    }
+}
+module.exports = cashnetJSON;
```
Use es6 export:

``` javascript
export default ClassName;
```

kyle-rader reviewed (changes requested):
Looking good so far!

kyle-rader commented on server/webhooks/registration-parser.js:
```diff
@@ -0,0 +1,84 @@
+            refArray[index - 1]['amount'] = cnj[`amount${index}`];
+            refArray[index - 1]['qty'] = cnj[`qty${index}`];
+        });
+        let refRE = /ref(\d{1,2})type(\d{1,2})/;
```
@dalingk I would make these variable names more explicit.

kyle-rader reviewed (commented):
See comment on var names around line 31

dalingk committed:
Updated registration-parser.js

Uses ES6 features
Uses destructuring to import lodash
Doesn't have babel-register stuff

dalingk committed:
Updated registration-parser.js

Allow unlimited items per order
Allow unlimited refs per order

dalingk committed:
Corrected valid tshirt detection

Orders for 3XL female t-shirts are now invalid
Invalid orders are now `false` instead of `null`

kyle-rader reviewed (approved):
@dalingk Great work with the updates.

There are some more review things I could add but for expediency I'm going to merge this.
One overall comment is to always prefer `'strings'` over `"strings"`.  It's just the general trend in JS.  You more often will be using a `"` literal in a string than a `'`.

kyle-rader commented:
@dalingk Once you see this is merged go ahead and delete the branch.  It is kosher for the branch creator to be the one that deletes it.
